halo2_proofs/examples/shuffle.rs

```diff
@@ -0,0 +1,351 @@
+use ff::BatchInvert;
+use halo2_proofs::{
+    arithmetic::{CurveAffine, FieldExt},
+    circuit::{floor_planner::V1, Layouter, Value},
+    dev::{metadata, FailureLocation, MockProver, VerifyFailure},
+    plonk::{Phase::*, *},
+    poly::{commitment::Params, Rotation},
+    transcript::{Blake2bRead, Blake2bWrite, Challenge255},
+};
+use pasta_curves::EqAffine;
+use rand_core::{OsRng, RngCore};
+use std::{
+    iter,
+    panic::{self, AssertUnwindSafe},
+};
+
+fn catch_unwind_silent<F: FnOnce() -> R, R>(f: F) -> std::thread::Result<R> {
+    let prev_hook = panic::take_hook();
+    panic::set_hook(Box::new(|_| {}));
+    let result = panic::catch_unwind(AssertUnwindSafe(f));
+    panic::set_hook(prev_hook);
+    result
+}
+
+fn rand_2d_array<F: FieldExt, R: RngCore, const W: usize, const H: usize>(
+    rng: &mut R,
+) -> [[F; H]; W] {
+    [(); W].map(|_| [(); H].map(|_| F::random(&mut *rng)))
+}
+
+fn shuffled<F: FieldExt, R: RngCore, const W: usize, const H: usize>(
+    original: [[F; H]; W],
+    rng: &mut R,
+) -> [[F; H]; W] {
+    let mut shuffled = original;
+
+    for row in (1..H).rev() {
+        let rand_row = (rng.next_u32() as usize) % row;
+        for column in shuffled.iter_mut() {
+            column.swap(row, rand_row);
+        }
+    }
+
+    shuffled
+}
+
+#[derive(Clone)]
+struct MyConfig<const W: usize> {
+    q_shuffle: Selector,
+    q_first: Selector,
+    q_last: Selector,
+    original: [Column<Advice>; W],
+    shuffled: [Column<Advice>; W],
+    theta: Challenge,
+    gamma: Challenge,
+    z: Column<Advice>,
+}
+
+impl<const W: usize> MyConfig<W> {
+    fn configure<F: FieldExt>(meta: &mut ConstraintSystem<F>) -> Self {
+        let [q_shuffle, q_first, q_last] = [(); 3].map(|_| meta.selector());
+        // Fist phase
+        let original = [(); W].map(|_| meta.advice_column_in(First));
+        let shuffled = [(); W].map(|_| meta.advice_column_in(First));
+        let [theta, gamma] = [(); 2].map(|_| meta.challenge_usable_after(First));
+        // Any phase skipped will cause a panic
+        assert!(catch_unwind_silent(|| meta.challenge_usable_after(Second)).is_err());
+        assert!(catch_unwind_silent(|| meta.advice_column_in(Third)).is_err());
```

Comment:
This would be appropriate for a test, but examples should just show the normal usage of the API, not demonstrate error conditions (especially error conditions we've chosen to expose as panics).